torch/csrc/jit/interpreter.cpp

```diff
@@ -479,72 +873,52 @@ struct InterpreterStateImpl {
     bool_data(function->bool_data),
     registers(function->register_size) {
   }
-  void runOneStage(
-    const std::vector<at::Tensor> & inputs,
-    std::vector<at::Tensor> & outputs) {
-      // std::cout << "running stage: " << current_stage << " of " << function->stages.size() << "\n";
-      // std::cout << *function->graph << "\n";
-      JIT_ASSERT(current_stage < function->stages.size());
-      auto & stage = function->stages[current_stage++];
-      JIT_ASSERT((int)inputs.size() == stage.inputs.size);
-      for(int i = 0; i < stage.inputs.size; i++) {
-        int reg = get(stage.inputs,i);
-        if(reg >= 0) { // otherwise this input is dead, and we do not store it to avoid holding the reference
-          registers.takeOwnership(reg, toRetainableShare(inputs[i]));
-        }
-        // std::cout << "registers[" << reg << "] = inputs[" << i << "](" << registers[reg] << ")\n";
-      }
-      for(auto & inst : stage.instructions) {
-        auto & inputs = inst.inputs.values;
-        for(int i = 0; i < inputs.size; i++) {
-          int reg = get(inputs,i);
-          input_buffer.push_back(registers[reg]);
-          // std::cout << "inputs[" << i << "] = registers[" << reg << "](" << registers[reg] << ")\n";
-        }
-        inst.callback(input_buffer, output_buffer);
-        for(int i = 0; i < inst.outputs.size; i++) {
+  void runOneStage(Stack & stack) {
+    // std::cout << "running stage: " << current_stage << " of " << function->stage_end.size() << "\n";
+    // std::cout << *function->graph << "\n";
+    // function->dump(std::cout);
+    size_t pc = current_pc;
+    size_t last = function->stage_end[current_stage];
+    auto & instructions = function->instructions;
+    while(pc < last) {
+        // std::cout << "executing " << pc << ": ";
+        // function->dumpInstruction(std::cout, pc);
+        // std::cout << "\n";
+        auto & inst = instructions[pc];
+        loadTensorsFromRegisters(inst.inputs, stack);
+        pc += 1 + inst.callback(stack);
+        for(int i = inst.outputs.size - 1; i >= 0; i--) {
           int reg = get(inst.outputs,i);
-          registers.takeOwnership(reg, std::move(output_buffer[i]));
-          // std::cout << "registers[" << reg << "] = outputs[" << i << "](" << registers[reg] << ")\n";
-        }
-        auto & frees = inst.inputs.free_flags;
-        for(int i = 0; i < frees.size; i++) {
-          if(get(frees,i)) {
-            registers.reset(get(inputs,i));
-          }
+          registers[reg] = pop(stack);
+          // std::cout << "pop reg[" << reg << "];\n" << registers[reg].pImpl << "\n";
         }
-        output_buffer.clear();
-        input_buffer.clear();
-      }
-      outputs.clear();
-      loadTensorsFromRegisters(stage.outputs, outputs);
+    }
+    current_pc = pc;
```

Comment:
Can't we just operate on `current_pc` all the time?

Reply:
I am worried about alias analysis not realizing `current_pc` can stay in a register because it is a member of a struct and we are calling indirect functions. Writing it this way proves to C++ that `pc` can be in a register.

Reply:
Yea I realized that later too. Makes sense, let's keep it this way :+1: